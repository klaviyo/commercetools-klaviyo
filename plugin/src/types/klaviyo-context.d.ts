import { KlaviyoService } from '../infrastructure/driven/klaviyo/KlaviyoService';
import { OrderMapper } from '../domain/shared/mappers/OrderMapper';
import { CustomerMapper } from '../domain/shared/mappers/CustomerMapper';
import { CategoryMapper } from '../domain/shared/mappers/CategoryMapper';
import { ProductMapper } from '../domain/shared/mappers/ProductMapper';
import { CtCustomerService } from '../infrastructure/driven/commercetools/CtCustomerService';
import { CtProductService } from '../infrastructure/driven/commercetools/CtProductService';
import { CtCategoryService } from '../infrastructure/driven/commercetools/CtCategoryService';
import { CtPaymentService } from '../infrastructure/driven/commercetools/CtPaymentService';
import { CtOrderService } from '../infrastructure/driven/commercetools/CtOrderService';

type Context = {
    klaviyoService: KlaviyoService;
    orderMapper: OrderMapper;
    customerMapper: CustomerMapper;
    categoryMapper: CategoryMapper;
    productMapper: ProductMapper;
    ctCustomerService: CtCustomerService;
    ctProductService: CtProductService;
    ctCategoryService: CtCategoryService;
    ctPaymentService: CtPaymentService;
    ctOrderService: CtOrderService;
};
