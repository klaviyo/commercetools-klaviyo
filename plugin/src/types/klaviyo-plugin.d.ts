type KlaviyoEvent = {
    type:
        | 'profileCreated'
        | 'profileUpdated'
        | 'profileResourceUpdated'
        | 'categoryCreated'
        | 'categoryDeleted'
        | 'categoryUpdated'
        | 'itemCreated'
        | 'itemUpdated'
        | 'itemDeleted'
        | 'variantCreated'
        | 'variantUpdated'
        | 'variantDeleted'
        | 'event';
    body: KlaviyoRequestType;
};

type KlaviyoRequestType =
    | ProfileRequest
    | EventRequest
    | CategoryRequest
    | CategoryDeletedRequest
    | ItemRequest
    | ItemVariantRequest
    | ItemJobRequest
    | ItemVariantJobRequest
    | ItemDeletedRequest;

type ProcessingResult = {
    status: 'OK' | '4xx';
};

type maxSizeJobArrayContainer = {
    [key: string]: KlaviyoEvent[];
};
