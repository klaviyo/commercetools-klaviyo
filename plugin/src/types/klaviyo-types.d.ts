type ProfileRequest = {
    data: ProfileType;
};

type ProfileType = {
    type: 'profile';
    id?: string;
    attributes: Profile;
    links?: any;
    relationships?: any;
    meta?: {
        identifiers: {
            external_id: string;
        };
    };
};

type Profile = {
    external_id?: string;
    anonymous_id?: string;
    email?: string;
    first_name?: string;
    last_name?: string;
    title?: string;
    phone_number?: string | null;
    organization?: string;
    image?: string | null;
    created?: string;
    updated?: string;
    last_event_date?: string;
    location?: KlaviyoLocation | null;
    properties?: any;
};

type KlaviyoLocation = {
    address1?: string | null;
    address2?: string | null;
    city?: string;
    country?: string;
    latitude?: string | null;
    longitude?: string | null;
    region?: string;
    zip?: string;
    timezone?: string | null;
};

type EventRequest = {
    data: EventType;
};

type EventType = {
    type: 'event';
    id?: string;
    attributes: {
        profile: KlaviyoEventProfile;
        metric: {
            name: string;
        };
        value: number;
        properties: any;
        unique_id: string;
        time: string;
    };
};

type KlaviyoEventProfile = {
    $email?: string;
    $id?: string;
    $first_name?: string;
    $last_name?: string;
    $phone_number?: string;
    [key: string]: unknown;
};

type CategoryRequest = {
    data: CategoryType;
};

type CategoryType = {
    type: 'catalog-category';
    id?: string;
    attributes: KlaviyoCategory;
    relationships?: KlaviyoRelationships;
};

type KlaviyoCategory = {
    external_id?: string;
    id?: string;
    name: string;
    integration_type?: string | '$custom';
    catalog_type?: string | '$default';
    updated?: string;
};

type CategoryDeletedRequest = {
    data: {
        id: string;
    };
};

type ItemRequest = {
    data: ItemType;
    variantJobRequests?: KlaviyoEvent[];
};

type ItemType = {
    type: 'catalog-item';
    id?: string;
    attributes: KlaviyoCatalogItem;
    relationships?: KlaviyoRelationships;
};

type KlaviyoCatalogItem = {
    external_id?: string;
    id?: string;
    integration_type?: string | '$custom';
    catalog_type?: string | '$default';
    title?: string;
    description?: string;
    url?: string;
    image_full_url?: string;
    published: boolean;
    price?: number;
    custom_metadata?: any;
};

type ItemVariantRequest = {
    data: ItemVariantType;
};

type ItemVariantType = {
    type: 'catalog-variant';
    id?: string;
    attributes: KlaviyoCatalogVariant;
    relationships?: KlaviyoRelationships;
};

type ItemDeletedRequest = {
    data: {
        id: string;
        deleteVariantsJob?: KlaviyoEvent;
    };
};

type KlaviyoCatalogVariant = {
    external_id?: string;
    id?: string;
    integration_type?: string | '$custom';
    catalog_type?: string | '$default';
    title?: string;
    description?: string;
    sku?: string;
    inventory_quantity?: number;
    inventory_policy?: 0 | 1 | 2;
    price?: number;
    url?: string;
    image_full_url?: string;
    published: boolean;
    custom_metadata?: any;
};

type KlaviyoRelationships = {
    [key: string]: {
        data: KlaviyoRelationshipData[]
    }
}

type KlaviyoRelationshipData = {
    type: 'catalog-item' | 'catalog-variant' | 'catalog-category';
    id: string;
}

type ItemJobRequest = {
    data: ItemJobType;
};

type ItemJobType = {
    type: 'catalog-item-bulk-create-job' | 'catalog-item-bulk-update-job';
    id?: string,
    attributes: {
        items: ItemType[];
    };
};

type ItemVariantJobRequest = {
    data: ItemVariantJobType;
};

type ItemVariantJobType = {
    type: 'catalog-variant-bulk-create-job' | 'catalog-variant-bulk-update-job' | 'catalog-variant-bulk-delete-job';
    id?: string,
    attributes: {
        variants: ItemVariantType[];
    };
};

type KlaviyoQueryResult<KlaviyoElement> = {
    data: KlaviyoElement[],
    links: {
        self: "string",
        first: "string",
        last: "string",
        prev: "string",
        next: "string"
    }
};
