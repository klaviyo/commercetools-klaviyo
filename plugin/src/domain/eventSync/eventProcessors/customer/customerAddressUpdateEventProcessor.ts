import { AbstractEventProcessor } from '../abstractEventProcessor';
import logger from '../../../../utils/log';
import {
    CustomerAddressAddedMessage,
    CustomerAddressChangedMessage,
    CustomerAddressRemovedMessage,
} from '@commercetools/platform-sdk/dist/declarations/src/generated/models/message';
import { getCTCustomerAddressForKlaviyo, getPhoneNumber } from './utils/CustomerAddressUtils';
import { mapCTAddressToKlaviyoLocation } from './mappers/CTAddressToKlaviyoLocationMapper';
import config from 'config';

export class CustomerAddressUpdateEventProcessor extends AbstractEventProcessor {
    isEventValid(): boolean {
        const message = this.ctMessage as unknown as
            | CustomerAddressAddedMessage
            | CustomerAddressRemovedMessage
            | CustomerAddressChangedMessage;
        return (
            message.resource.typeId === 'customer' &&
            Boolean(config.get<string[]>('customer.messages.addressChanged').includes(message.type))
        );
    }

    async generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const message = this.ctMessage as unknown as
            | CustomerAddressAddedMessage
            | CustomerAddressRemovedMessage
            | CustomerAddressChangedMessage;
        logger.info(`processing CT ${message.type} message`);
        const customer = await this.context.ctCustomerService.getCustomerProfile(message.resource.id);
        const address = getCTCustomerAddressForKlaviyo(customer);

        const body: ProfileRequest = {
            data: {
                type: 'profile',
                attributes: {
                    phone_number: getPhoneNumber(address),
                    location: mapCTAddressToKlaviyoLocation(address),
                },
                meta: {
                    identifiers: {
                        external_id: message.resource.id,
                    },
                },
            },
        };
        return [
            {
                body: body,
                type: 'profileUpdated',
            },
        ];
    }
}
