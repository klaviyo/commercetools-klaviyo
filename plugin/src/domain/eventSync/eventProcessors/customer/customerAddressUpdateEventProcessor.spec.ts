import { MessageDeliveryPayload } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/subscription';
import { DeepMockProxy, mockDeep } from 'jest-mock-extended';
import { expect as exp } from 'chai';
import { CustomerAddressUpdateEventProcessor } from './customerAddressUpdateEventProcessor';
import { getSampleCustomerAddressUpdateMessage } from '../../../../test/testData/ctCustomerMessages';
import { getSampleCustomerApiResponse } from '../../../../test/testData/ctCustomerApi';
import { Context } from "../../../../types/klaviyo-context";

const ctCustomerServiceMock = {
    getCustomerProfile: jest.fn(),
}

jest.mock('../../../../infrastructure/driven/commercetools/DefaultCtCustomerService', () => ({
    DefaultCtCustomerService: function() {
        return ctCustomerServiceMock;
    }
}));

jest.mock('../../../../infrastructure/driven/commercetools/ctService');
const contextMock: DeepMockProxy<Context> = mockDeep<Context>();
describe('CustomerAddressUpdateEventProcessor > isEventValid', () => {
    it.each`
        type
        ${'CustomerAddressAdded'}
        ${'CustomerAddressRemoved'}
        ${'CustomerAddressChanged'}
    `('should return valid when the message type is $type', ({ type }) => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'customer' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: type });

        const event = CustomerAddressUpdateEventProcessor.instance(ctMessageMock, contextMock);

        exp(event.isEventValid()).to.be.true;
    });

    it.each`
        resource      | type
        ${'invalid'}  | ${'CustomerAddressUpdate'}
        ${'customer'} | ${'invalid'}
    `('should return invalid when is not a valid customer address update message', ({ resource, type }) => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: resource } });
        Object.defineProperty(ctMessageMock, 'type', { value: type });

        const event = CustomerAddressUpdateEventProcessor.instance(ctMessageMock, contextMock);

        exp(event.isEventValid()).to.be.false;
    });
});

describe('CustomerAddressUpdateEventProcessor > generateKlaviyoEvent', () => {
    it('should generate the klaviyo event when the customer address update event is valid', async () => {
        const inputMessage = getSampleCustomerAddressUpdateMessage();
        const message = inputMessage as unknown as MessageDeliveryPayload;
        const event = CustomerAddressUpdateEventProcessor.instance(message, contextMock);
        contextMock.ctCustomerService.getCustomerProfile.mockResolvedValue(getSampleCustomerApiResponse());

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(klaviyoEvent[0].body).toMatchSnapshot();
    });

    it("should set the profile location to null in Klaviyo if CT doesn't return and address", async () => {
        const inputMessage = getSampleCustomerAddressUpdateMessage();
        const message = inputMessage as unknown as MessageDeliveryPayload;
        const event = CustomerAddressUpdateEventProcessor.instance(message, contextMock);
        contextMock.ctCustomerService.getCustomerProfile.mockResolvedValue(getSampleCustomerApiResponse([]));

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(klaviyoEvent[0].body).toMatchSnapshot();
    });
});
