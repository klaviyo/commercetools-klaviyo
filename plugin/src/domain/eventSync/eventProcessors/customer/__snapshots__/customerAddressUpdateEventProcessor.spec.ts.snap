// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerAddressUpdateEventProcessor > generateKlaviyoEvent should generate the klaviyo event when the customer address update event is valid 1`] = `
{
  "data": {
    "attributes": {
      "location": {
        "address1": "12, First Street",
        "address2": "",
        "city": "Brookefort",
        "country": "NL",
        "region": undefined,
        "zip": "12345",
      },
      "phone_number": "+393234567892",
    },
    "meta": {
      "identifiers": {
        "external_id": "77d8e6f4-2dee-4ac9-bd21-4035b6ba76d8",
      },
    },
    "type": "profile",
  },
}
`;

exports[`CustomerAddressUpdateEventProcessor > generateKlaviyoEvent should set the profile location to null in Klaviyo if CT doesn't return and address 1`] = `
{
  "data": {
    "attributes": {
      "location": null,
      "phone_number": null,
    },
    "meta": {
      "identifiers": {
        "external_id": "77d8e6f4-2dee-4ac9-bd21-4035b6ba76d8",
      },
    },
    "type": "profile",
  },
}
`;
